### PR TITLE
Feature metric manager

### PR DESCRIPTION
**Pull request by PososikTeam**
Create MetricManager class with 2 classes MetricMemoryBank for Retrieval and MetricWrapper - wrapper for use TorchMetrics. Also write some tests.

### CODE REVIEW

File: src/metrics/metric_manager.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            phase_memory_bank = self.memory_bank[phase].compute()
+            
+        log = {}
+        for name, metric in self.metrics[phase].items():
+            if hasattr(metric, 'require_memory_bank'):
+                extension_name, metric_value = metric.compute(phase_memory_bank)
+            else:
+                extension_name, metric_value = metric.compute()
+            
+            metric_value_shape = list(metric_value.shape)
```
Comment (VladVin) [inline src/metrics/metric_manager.py:71]: Это не нужно

File: tests/metrics/data_generator.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    shape: List[int]
+
+
+class FakeDataGenerator:
+    def __init__(self, fake_data_list: List[FakeData], fake_data_sizes: List[int], bs: int = 4):
+        self.fake_data_list = fake_data_list
+        self.fake_data_sizes = fake_data_sizes
+        self.bs = bs
+        
+    def _generate_fake_data(self, fake_data: FakeData):
```
Comment (VladVin) [inline tests/metrics/data_generator.py:19]: Pass shape only

File: tests/metrics/data_generator.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from typing import List
+import torch
+
+from dataclasses import dataclass
+
+
+@dataclass
+class FakeData:
+    name: str
+    shape: List[int]
```
Comment (VladVin) [inline tests/metrics/data_generator.py:10]: Add `num_repeats` field

File: tests/metrics/test_metric_manager.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+    def update(self, predict: torch.Tensor, target: torch.Tensor):
+        self.lol_sum += 1
+
+    def compute(self):
+        return self.lol_sum
+
+    
+@METRICS.register_class
+class MocRequoreMemoryBlockMetric(Metric):
```
Comment (VladVin) [inline tests/metrics/test_metric_manager.py:45]: `MockRequireMemoryBlockMetric`

File: src/metrics/metric_utils.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        return obj_name2memory
+
+
+class MetricWrapper:
+    #!TODO add documentation
+    def __init__(self, metric: Metric, target_fields: Dict, name: str = None, phases: List[str] = ['train', 'valid']):
+        self.metric = metric
+        self.name = name if name is not None else type(metric).__name__
+        self.phases = phases
+        self.require_memory_bank = True if hasattr(metric, 'require_memory_bank') else False
```
Comment (VladVin) [inline src/metrics/metric_utils.py:36]: It should be property

File: src/metrics/metric_utils.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                self.__getattribute__(name).append(value)
+
+    def compute(self):
+        obj_name2memory = {}
+        for name in self.tracked_obj_names:
+            obj_name2memory[name] = torch.cat(self.__getattribute__(name), 0)
+        return obj_name2memory
+
+
+class MetricWrapper:
```
Comment (VladVin) [inline src/metrics/metric_utils.py:30]: Think of using decorators

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/metrics/metric_manager.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            Metric list as nn.ModuleList for current phase. 
+        """
+        # create added_metric_names set
+        added_metric_names = set()
+        metrics = []
+        for metric_params in params:
+            if phase not in metric_params.phases:
+                continue
+            metric = METRICS.get(metric_params.class_name)(**metric_params.metric_params)
+            target_fields = metric_params.target_fields
```
Comment (VladislavPatrushev) [inline src/metrics/metric_manager.py:98]: May be should remove target_fields and add target and prediction

File: src/metrics/metric_manager.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        for metric_with_utils in self.phase2metrics[phase]:
+            metric_value = metric_with_utils.compute()
+            if isinstance(metric_value, dict):
+                metric_value = list(metric_value.values())[0]
+            
+            if len(metric_value.shape) != 0:
+                raise ValueError(f'{metric_with_utils.name} must compute float value, '
+                                f'not torch tensor with shap {metric_value.shape}.')
+            
+            log[f'{phase}/{metric_with_utils.name}'] = metric_value
```
Comment (VladislavPatrushev) [inline src/metrics/metric_manager.py:156]: Too long and complex key f'{phase}/{metric_with_utils.name}'.
Maybe:
metric_key = f'{phase}/{metric_with_utils.name}'
log[metric_key] = metric_value

File: src/metrics/metric_manager.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        self.class_name = class_name
+        self.target_fields = target_fields
+        self.phases = phases if phases is not None else ['train', 'valid', 'test']
+        self.name = name
+        self.metric_params = metric_params
+
+
+class MetricWithUtils(nn.Module):
+    """
+    Union class for metric and metric utils parameters
```
Comment (VladislavPatrushev) [inline src/metrics/metric_manager.py:33]: Add dot

File: src/metrics/metric_manager.py (hunk 7 above)
Comment (PososikTeam) [reply src/metrics/metric_manager.py:98]: I think exist metric with more than one input argument. Moreover if we do that it force me to rewrite all torchmetrics metrics.

File: src/metrics/metric_manager.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    # model use phases
+    phases = [Phase.TRAIN, Phase.VALID, Phase.TEST, Phase.PREDICT]
+
+    def __init__(self, params: List[MetricParams]):
+        super().__init__()
+        # Change list to set in phases
+        for param in params:
+            param.phases = set(param.phases)
+
+        phase2metrics = {phase: {} for phase in self.phases}
```
Comment (VladVin) [inline src/metrics/metric_manager.py:103]: I think you don't need to create dict in such a way. Do it simpler:
```
self.__phase2metrics = {}
for phase in Phase:
  ...
```

File: src/metrics/metric_manager.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    VALID = 'valid'
+    TEST = 'test'
+    PREDICT = 'predict'
+
+phase_mapping = {
+    'train': Phase.TRAIN,
+    'valid': Phase.VALID,
+    'test': Phase.TEST,
+    'predict': Phase.PREDICT,
+    'Phase.TRAIN': Phase.TRAIN,
```
Comment (VladVin) [inline src/metrics/metric_manager.py:23]: I think you don't need these `Phase.*` options

**Review by VladVin — CHANGES_REQUESTED** [review]

File: tests/metrics/metric_manager/test_metric_manager.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+    def update(self, predict: torch.Tensor, target: torch.Tensor):
+        self.lol_sum += 1
+
+    def compute(self):
+        return self.lol_sum
+
+
+@METRICS.register_class
+class MockSumMetric(Metric):
```
Comment (VladVin) [inline tests/metrics/metric_manager/test_metric_manager.py:27]: Repeated class

File: tests/metrics/metric_manager/test_metric_manager.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        super().__init__(**kwargs)
+
+    def update(self, predict: torch.Tensor, target: torch.Tensor):
+        return
+
+    def compute(self, memory_block):
+        return torch.tensor([1, 2])
+
+
+def run_metric_manager(class_names: List[str], names: List[str], \
```
Comment (VladVin) [inline tests/metrics/metric_manager/test_metric_manager.py:64]: Maybe to call arguments as they should be called? `metric_names`, `log_names`, `mappings`

File: tests/metrics/metric_manager/test_metric_manager.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+                       target_fields: List[Dict], data_generator: FakeDataGenerator):
+    if len(class_names) != len(names) or len(names) != len(target_fields):
+        raise print('Not correct Test! Input params not same length.')
```
Comment (VladVin) [inline tests/metrics/metric_manager/test_metric_manager.py:67]: Fix exception type

File: tests/metrics/metric_manager/test_metric_manager.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+# generate data
+embedding_data = FakeData(name='embedding', shape = [512], num_repeats=5)
+target_data = FakeData(name='target', shape=[10], num_repeats=5)
+embedding_fake_data = [embedding_data, target_data]
+data_generator = FakeDataGenerator(embedding_fake_data)
+target_fields = dict(predict='embedding', target='target')
+
+
+class MetricManagerTest(unittest.TestCase):
+    def test(self):
```
Comment (VladVin) [inline tests/metrics/metric_manager/test_metric_manager.py:101]: Rename according to template: `test_something_when_something_happens`

File: tests/metrics/metric_manager/test_metric_manager_ddp.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+    def __len__(self):
+        return len(self.labels)
+
+    def __getitem__(self, item):
+        return torch.rand(INPUT_DATA_SHAPE), torch.tensor(self.predicts[item]), torch.tensor(self.labels[item])
+
+
+@METRICS.register_class
+class MetricMemoryBlock(Metric):
```
Comment (VladVin) [inline tests/metrics/metric_manager/test_metric_manager_ddp.py:47]: Not exists anymore

File: tests/metrics/metric_manager/test_metric_manager_ddp.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+
+class Model(LightningModule):
+    def __init__(self, metric_params: List[MetricParams]):
+        super().__init__()
+        self.l1 = torch.nn.Linear(INPUT_DATA_SHAPE, uniq_label_count)
+        self.metric_manager = MetricManager(metric_params)
+
+    def forward(self, x):
+        return torch.relu(self.l1(x.view(x.size(0), -1)))
```
Comment (VladVin) [inline tests/metrics/metric_manager/test_metric_manager_ddp.py:66]: Why not to mock the model and training steps?

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/metrics/metric_manager.py [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+    'Phase.TEST': Phase.TEST,
+    'Phase.PREDICT': Phase.PREDICT,
+}
+
+@dataclass
+class MetricParams:
+    name: str
+    mapping: Dict[str, str]
+    log_name: str = None
+    params: Dict = field(default_factory=dict)
```
[the remark this answers is not in the document]
Comment (PososikTeam) [reply src/metrics/metric_manager.py:34]: Because it call torchmetrics Metric constructor with `**params` and if `params is None` then line like `Accuracy(**params)` will throws an error.

File: src/metrics/metric_manager.py [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+        """
+        if phase not in self.phases:
+            raise ValueError(f'Incorrect epoch setting. '
+                             f'Please choose one of enum value {self.phases}')
+            
+        log = {}
+        for metric_with_utils in self.__phase2metrics[phase]:
+            metric_value = metric_with_utils.compute()
+            if isinstance(metric_value, dict):
+                metric_value = list(metric_value.values())[0]
```
[the remark this answers is not in the document]
Comment (PososikTeam) [reply src/metrics/metric_manager.py:178]: Torchmetrics detection mAP return dictionary with metrics https://torchmetrics.readthedocs.io/en/stable/detection/mean_average_precision.html.
So i include dict type as supported type.

File: src/metrics/metric_manager.py [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+            
+            if len(metric_value.shape) != 0:
+                raise ValueError(f'{metric_with_utils.log_name} must compute float value, '
+                                f'not torch tensor with shap {metric_value.shape}.')
```
[the remark this answers is not in the document]
Comment (PososikTeam) [reply src/metrics/metric_manager.py:182]: Ok, now we must decide what is good type, i think it's just a number, because we return logs with string value. And some metrics such like Confusion Matrix would be bad to write as string in PL logger. 
What do you think?

File: tests/metrics/metric_manager/test_metric_manager_ddp.py (hunk 16 above)
Comment (PososikTeam) [reply tests/metrics/metric_manager/test_metric_manager_ddp.py:47]: It's need to test.

File: tests/metrics/metric_manager/test_metric_manager_ddp.py (hunk 17 above)
Comment (PososikTeam) [reply tests/metrics/metric_manager/test_metric_manager_ddp.py:66]: How to mock PL Module?

File: src/metrics/metric_manager.py [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+
+        Args:
+            phase: Phase Enum.
+        """
+        args = list(args)
+
+        for metric_with_utils in self.__phase2metrics[phase.name]:
+            targeted_kwargs = self.map_arguments(metric_with_utils.mapping, kwargs)
+            if targeted_kwargs:
+                # may be we only need to update because forward makes synchronization between processes
```
Comment (VladVin) [inline src/metrics/metric_manager.py:163]:
```suggestion
```

File: src/metrics/metric_manager.py [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+        Args:
+            mapping: Dictionary for mapping Metric forward input keys with Task output dictionary keys.
+            task_output: Output after task forward pass.
+
+        Returns:
+            metric_input: Metric input dictionary like **kwargs for metric forward pass.
+        """
+        metric_input = {}
+        for metric_target, metric_source in mapping.items():
+            if metric_source in task_output:
```
Comment (VladVin) [inline src/metrics/metric_manager.py:222]: Raise error if at least one key is missing as it's done in Constructor

File: src/metrics/metric_manager.py [hunk 23]
```diff
@@ -0,0 +1,230 @@
...
+            if phase not in metric_params.phases:
+                continue
+            metric = METRICS.get(metric_params.name)(**metric_params.params)
+            mapping = metric_params.mapping
+            prefix = '' if metric_params.prefix is None else metric_params.prefix + '_'
+            log_name = prefix + metric_params.name
+            if log_name in added_log_names:
+                # If prefix not set.
+                if metric_params.prefix is None:
+                    raise ValueError('Has a two identical metrics. Please, set in config file '
```
Comment (VladVin) [inline src/metrics/metric_manager.py:138]: "a" is not used in front of plural nouns in English. Also, I think you don't need defferent behaviours depending on prefix being set. Just use one:
`Got two metrics with identical names: {log_name}. Please, set differet prefixes for identical metrics in the config file`

File: tests/metrics/metric_manager/test_metric_manager.py (hunk 15 above)
Comment (VladVin) [reply tests/metrics/metric_manager/test_metric_manager.py]: Lol, I meant to replaces <something> and <something_happens>. Check my example: `def test_no_weight_decay_in_optimizers_when_module_contains_1d_tensor(self):`

File: src/metrics/metrics_manager.py [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+        """Update states of all metrics on phase loop.
+
+        Args:
+            phase: Phase Enum.
+        """
+        args = list(args)
+
+        for metric_with_utils in self.__phase2metrics[phase.name]:
+            targeted_kwargs = self.map_arguments(metric_with_utils.mapping, kwargs)
+            if targeted_kwargs:
```
Comment (VladVin) [inline src/metrics/metrics_manager.py:159]: Not needed anymore

**Review by VladVin — CHANGES_REQUESTED** [review]
